### ISSUE TITLE
Linkerd stuck talking to stale address after headless service rollout

### ISSUE THREAD

**Issue by adleong** [linkerd/linkerd2#5209]
### What is the issue?

When a meshed pod is sending requests to a headless service and the contents of that service change (e.g. from rolling the deployment), the Linkerd proxy in the client pod will be stuck attempting to connect to a stale endpoint.

This happens because when the client resolves the headless service, it will attempt to connect to an endpoint of that service (rather than a service clusterip).  The Linkerd proxy will forward the connection to that endpoint, but when the target deployment is restarted and that endpoint goes away, the Linkerd proxy does not break the connection from the client container.

### How can it be reproduced?

* Have a meshed pod send traffic to a headless service
* Restart the deployment which backs the service

### Logs, error output, etc

The client's Linkerd proxy logs:

```
[346867.852718s]  WARN ThreadId(01) outbound:accept{peer.addr=XXX.XXX.XXX.XXX:XXXX target.addr=XXX.XXX.XXX.XXX:XXXX}: linkerd2_app_core::errors: Failed to proxy request: error trying to connect: Connection refused (os error 111)
```

where the target.addr is a stale endpoint which no longer exists.

### Environment

- Linkerd version: edge-20.11.1

**Comment by olix0r** [comment]
https://github.com/linkerd/linkerd2-proxy/pull/736 and https://github.com/linkerd/linkerd2-proxy-api/pull/51 sketch out a proxy implementation that supports an `endpoint` response on profile resolutions, intended to be set when the resolved IP is known to be a k8s pod. When this field is set, further discovery is bypassed (including the balancer and traffic split router). When the resolved endpoint is neither a service nor a known pod, we should skip setting a `dst_override` so that these stacks may be bypassed.

When forwarding to a singular endpoint and not a service, we handle `io::Errors` on requests specially so that the entire service becomes defunct.

We should try to implement this resolution scheme in the destination controller to validate that this resolves this issue.

**Comment by olix0r** [comment]
Fixed by #5227 & #5253
